src/JsonRpcEngine.ts

```diff
@@ -0,0 +1,350 @@
+import SafeEventEmitter from '@metamask/safe-event-emitter';
+import { errorCodes, EthereumRpcError, serializeError } from 'eth-rpc-errors';
+
+type Maybe<T> = T extends Record<string, unknown> ? Partial<T> : T | null | undefined;
+
+export type Json =
+  | boolean
+  | number
+  | string
+  | null
+  | { [property: string]: Json}
+  | Json[];
+
+/**
+ * A String specifying the version of the JSON-RPC protocol.
+ * MUST be exactly "2.0".
+ */
+export type JsonRpcVersion = '2.0';
+
+/**
+ * An identifier established by the Client that MUST contain a String, Number,
+ * or NULL value if included. If it is not included it is assumed to be a
+ * notification. The value SHOULD normally not be Null and Numbers SHOULD
+ * NOT contain fractional parts.
+ */
+export type JsonRpcId = number | string | void;
+
+export interface JsonRpcError {
+  code: number;
+  message: string;
+  data?: unknown;
+  stack?: string;
+}
+
+export interface JsonRpcRequest<T> {
+  jsonrpc: JsonRpcVersion;
+  method: string;
+  id: JsonRpcId;
+  params?: T;
+}
+
+export interface JsonRpcNotification<T> {
+  jsonrpc: JsonRpcVersion;
+  method: string;
+  params?: T;
+}
+
+interface JsonRpcResponseBase {
+  jsonrpc: JsonRpcVersion;
+  id: JsonRpcId;
+}
+
+export interface JsonRpcSuccess<T> extends JsonRpcResponseBase {
+  result: Maybe<T>;
+}
+
+export interface JsonRpcFailure extends JsonRpcResponseBase {
+  error: JsonRpcError;
+}
+
+export type JsonRpcResponse<T> = JsonRpcSuccess<T> | JsonRpcFailure;
+
+export type JsonRpcEngineCallbackError = Error | JsonRpcError | null;
+
+export type JsonRpcEngineReturnHandler = (
+  done: (error?: JsonRpcEngineCallbackError) => void
+) => void;
+
+export type JsonRpcEngineNextCallback = (
+  returnHandlerCallback?: JsonRpcEngineReturnHandler
+) => void;
+
+export type JsonRpcEngineEndCallback = (
+  error?: JsonRpcEngineCallbackError
+) => void;
+
+export type JsonRpcMiddleware = (
+  req: JsonRpcRequest<unknown>,
+  res: JsonRpcResponse<unknown>,
+  next: JsonRpcEngineNextCallback,
+  end: JsonRpcEngineEndCallback
+) => void;
```

Comment:
```suggestion
export type JsonRpcMiddleware<T, U> = (
  req: JsonRpcRequest<T>,
  res: JsonRpcResponse<U>,
  next: JsonRpcEngineNextCallback,
  end: JsonRpcEngineEndCallback
) => void;
```

Reply:
I implemented this, although still passed `<unknown, unknown>` to `asMiddleware`, since that's converting an entire engine into a middleware, and it will process any number of different request and response types. Not sure if it's worth doing anything differently with respect to that. 4f9a95e

Reply:
Actually: 9bbb4ae